iceoryx_hoofs/test/moduletests/error_reporting/module_a/errors.hpp

```diff
@@ -0,0 +1,100 @@
+#ifndef IOX_HOOFS_MODULETESTS_ERROR_REPORTING_MODULE_A_ERRORS_HPP
+#define IOX_HOOFS_MODULETESTS_ERROR_REPORTING_MODULE_A_ERRORS_HPP
+
+// ***
+// *** TO BE IMPLEMENTED BY CLIENT - part of any module
+// ***
+
+#include "iceoryx_hoofs/error_reporting/error.hpp"
+
+namespace module_a
+{
+namespace errors
+{
+
+using error_code_t = iox::err::error_code_t;
+using module_id_t = iox::err::module_id_t;
+
+enum class ErrorCode : error_code_t
+{
+    Unknown = 42,
+    OutOfMemory = 73,
+    OutOfBounds = 21
+};
+
+class Error
+{
+  public:
+    explicit Error(ErrorCode code = ErrorCode::Unknown)
+        : m_code(code)
+    {
+    }
+
+    static constexpr module_id_t module()
+    {
+        return MODULE_ID;
+    }
+
+    error_code_t code() const
+    {
+        return static_cast<error_code_t>(m_code);
+    }
+
+    const char* name() const
+    {
+        return errorNames[code()];
+    }
+
+    static constexpr module_id_t MODULE_ID = 1;
+
+  protected:
+    ErrorCode m_code;
+
+    static constexpr const char* errorNames[] = {"Unknown", "OutOfMemory", "OutOfBounds"};
+};
+
+// could be wrapped by a result/optional monadic type
+// could also be implemented without inheritence
+class OutOfBoundsError : public Error
+{
+  public:
+    OutOfBoundsError()
+        : Error(ErrorCode::OutOfBounds)
+    {
+    }
+
+    void* details()
+    {
+        return m_details;
+    }
+
+  private:
+    // more infos if available
+    void* m_details{nullptr};
+};
+
+} // namespace errors
+} // namespace module_a
+
+namespace iox
+{
+namespace err
+{
+
+// This definition must exist in this namespace for overload resolution.
+// Each module must use a unqiue error enum, e.g. by namespace.
+inline module_a::errors::Error toError(module_a::errors::ErrorCode code)
+{
+    return module_a::errors::Error(code);
+}
```

Comment:
The default implementation will returns the same object (maybe there should be no default for safety?). This works, as long as the type returned implements `code` and `module` (and maybe later some other functions).

This is to make it work out of the box for any type conforming to the interface.

This is not the case for an enum like here, hence it would break at compile time. This specialization takes care of it by defining how to construct an error from a code. I see no issue here, except that the implicit default identity behavior may be undesirable (a matter of preference to me).